### PR TITLE
Arrow time series fixes

### PR DESCRIPTION
**Pull request by agibsonccc**
## What changes were proposed in this pull request?
Fixes for arrow's time series indexing and the local executioner.
(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

Please review
https://github.com/deeplearning4j/deeplearning4j/blob/master/CONTRIBUTING.md before opening a pull request.

### CODE REVIEW

File: datavec-api/src/main/java/org/datavec/api/transform/schema/conversion/TypeConversion.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        return Double.parseDouble(o);
+    }
+
+    public double convertDouble(Object o) {
+        if(o instanceof Writable) {
+            Writable writable = (Writable) o;
+            return convertDouble(writable);
+        }
+        else {
+            return convertDouble(o.toString());
```
Comment (AlexDBlack) [inline datavec-api/src/main/java/org/datavec/api/transform/schema/conversion/TypeConversion.java]: Minor optimization here, and all the convertX(Object) methods (depending on how we want to use this): I'd add an ```else if( o instanceof Number )```

File: datavec-arrow/src/main/java/org/datavec/arrow/ArrowConverter.java [hunk 2]
```diff
@@ -80,35 +105,53 @@ public static INDArray toArray(ArrowWritableRecordBatch arrowWritableRecordBatch
...
-                    arr.putColumn(i,Nd4j.create(doubleBuffer, new int[] {cols,1}));
-
-                    break;
-                case Long:
-                    BigIntVector bigIntVector = (BigIntVector) columnVectors.get(i);
-                    DataBuffer longBuffer = Nd4j.createBuffer(bigIntVector.getDataBuffer().unwrap().nioBuffer(), DataBuffer.Type.LONG,cols);
-                    arr.putColumn(i,Nd4j.create(longBuffer, new int[] {cols,1}));
+                case DOUBLE:
+                    arr.putColumn(i,put.convertToDoubles());
                     break;
```
Comment (AlexDBlack) [inline datavec-arrow/src/main/java/org/datavec/arrow/ArrowConverter.java]: Minor point: add a default case + exception (mainly so when we have mixed datatypes in ND4J in the future, the error/issue is very obvious)
Edit: or better yet: maybe just make the default behaviour for all other data types be convertToDoubles().

File: datavec-arrow/src/main/java/org/datavec/arrow/ArrowConverter.java [hunk 3]
```diff
@@ -744,14 +952,64 @@ public static Writable fromEntry(int item,FieldVector from,ColumnType columnType
...
+        }
+
+        throw new IllegalArgumentException("Illegal vector type for int " + fieldVector.getClass().getName());
+    }
+
+    private static double getDoubleFromFieldVector(int row,FieldVector fieldVector) {
+        if(fieldVector instanceof Float8Vector) {
+            Float8Vector uInt4Vector = (Float8Vector) fieldVector;
+            return uInt4Vector.get(row);
+        }
```
Comment (AlexDBlack) [inline datavec-arrow/src/main/java/org/datavec/arrow/ArrowConverter.java]: Not 100% sure how this is used, but wouldn't Float4Vector be a valid ```else if``` condition here?

File: datavec-arrow/src/main/java/org/datavec/arrow/recordreader/ArrowWritableRecordBatch.java [hunk 4]
```diff
@@ -20,14 +19,24 @@
  */
 @Data
 @AllArgsConstructor
-public class ArrowWritableRecordBatch extends AbstractWritableRecordBatch implements List<List<Writable>>,Closeable {
+public class ArrowWritableRecordBatch implements List<List<Writable>>,Closeable {
```
Comment (AlexDBlack) [inline datavec-arrow/src/main/java/org/datavec/arrow/recordreader/ArrowWritableRecordBatch.java]: Why change this back and re-add dozens of unnecessary methods?
The way I had it was much cleaner...

File: datavec-arrow/src/main/java/org/datavec/arrow/recordreader/ArrowWritableRecordBatch.java [hunk 5]
```diff
@@ -78,6 +191,59 @@ public void close() throws IOException {
             arrowRecordBatch.close();
         if(vectorLoader != null)
             vectorLoader.close();
+
+
+    }
+
+
+    private class ArrowListIterator implements ListIterator<List<Writable>> {
```
Comment (AlexDBlack) [inline datavec-arrow/src/main/java/org/datavec/arrow/recordreader/ArrowWritableRecordBatch.java]: As above - why re-introduce ArrowListIterator?

File: datavec-arrow/src/main/java/org/datavec/arrow/recordreader/ArrowWritableRecordTimeSeriesBatch.java [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.*;
+
+/**
+ *
+ */
+@Data
+@AllArgsConstructor
+public class ArrowWritableRecordTimeSeriesBatch implements List<List<List<Writable>>>,Closeable {
```
Comment (AlexDBlack) [inline datavec-arrow/src/main/java/org/datavec/arrow/recordreader/ArrowWritableRecordTimeSeriesBatch.java]: If I was implementing this: I'd implement a sequence version of the following, and extend it: https://github.com/deeplearning4j/DataVec/blob/master/datavec-api/src/main/java/org/datavec/api/writable/batch/AbstractWritableRecordBatch.java

File: datavec-arrow/src/test/java/org/datavec/arrow/ArrowConverterTest.java [hunk 7]
```diff
@@ -231,6 +299,11 @@ public void testRecordReaderMetaData() throws Exception {
     }
 
 
+    @Test
+    public void testArrowWritablesTimeSeries() {
```
Comment (AlexDBlack) [inline datavec-arrow/src/test/java/org/datavec/arrow/ArrowConverterTest.java]: Empty test

File: datavec-spark-inference-parent/datavec-spark-inference-model/src/main/java/org/datavec/spark/transform/CSVSparkTransform.java [hunk 8]
```diff
@@ -162,12 +183,14 @@ public Base64NDArrayBody transformSequenceArray(SequenceBatchCSVRecord batchCSVR
...
+        INDArray arr = RecordConverter.toTensor(arrowWritableRecordBatch);
         try {
-            return new Base64NDArrayBody(Nd4jBase64
-                    .base64String(RecordConverter
-                            .toTensor(transformProcess.executeToSequenceBatch(
-                                    transformProcess.transformRawStringsToInputSequence(
-                                            singleCsvRecord.getRecordsAsString())))));
+            return new Base64NDArrayBody(Nd4jBase64.base64String(arr));
         } catch (IOException e) {
             e.printStackTrace();
```
Comment (AlexDBlack) [inline datavec-spark-inference-parent/datavec-spark-inference-model/src/main/java/org/datavec/spark/transform/CSVSparkTransform.java]: Another stack trace print here (not added in this PR though)

File: datavec-spark-inference-parent/datavec-spark-inference-server/pom.xml [hunk 9]
```diff
@@ -94,6 +94,11 @@
             <artifactId>datavec-data-image</artifactId>
             <version>${parent.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.datavec</groupId>
+            <artifactId>datavec-local</artifactId>
+            <version>${nd4j.version}</version>
```
Comment (AlexDBlack) [inline datavec-spark-inference-parent/datavec-spark-inference-server/pom.xml]: To be really pedantic: should be project.version or something?

File: datavec-arrow/src/main/java/org/datavec/arrow/recordreader/ArrowWritableRecordTimeSeriesBatch.java [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+     * An index in to an individual
+     * {@link ArrowRecordBatch}
+     * @param list the list of field vectors to use
+     * @param schema the schema to use
+     */
+    public ArrowWritableRecordTimeSeriesBatch(List<FieldVector> list, Schema schema,int timeSeriesStride) {
+        this.list = list;
+        this.schema = schema;
+        //each column should have same number of rows
+        this.timeSeriesStride = timeSeriesStride;
```
Comment (AlexDBlack) [inline datavec-arrow/src/main/java/org/datavec/arrow/recordreader/ArrowWritableRecordTimeSeriesBatch.java]: ArrowWritableRecordTimeSeriesBatch seems to only support sequences that have exactly the same number of time steps?
If so - that's a huge restriction, and makes ArrowWritableRecordTimeSeriesBatch not useful/suitable for general purpose use in DataVec.

File: datavec-arrow/src/main/java/org/datavec/arrow/recordreader/ArrowWritableRecordTimeSeriesBatch.java [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+
+    @Override
+    public void clear() {
+
+    }
+
+    @Override
+    public List<List<Writable>> get(int i) {
+        int timeStepLength = timeSeriesStride;
+        int offset = (timeStepLength * i) / timeStepLength;
```
Comment (AlexDBlack) [inline datavec-arrow/src/main/java/org/datavec/arrow/recordreader/ArrowWritableRecordTimeSeriesBatch.java]: This offset indexing isn't correct.

File: datavec-arrow/src/main/java/org/datavec/arrow/recordreader/ArrowWritableRecordTimeSeriesBatch.java (hunk 10 above)
Comment (agibsonccc) [reply datavec-arrow/src/main/java/org/datavec/arrow/recordreader/ArrowWritableRecordTimeSeriesBatch.java]: The local executioner has the proper changes for now.

File: datavec-arrow/src/test/java/org/datavec/arrow/ArrowConverterTest.java (hunk 7 above)
Comment (agibsonccc) [reply datavec-arrow/src/test/java/org/datavec/arrow/ArrowConverterTest.java]: Removed.

File: datavec-spark-inference-parent/datavec-spark-inference-server/pom.xml (hunk 9 above)
Comment (agibsonccc) [reply datavec-spark-inference-parent/datavec-spark-inference-server/pom.xml]: That won't work for this due to the spark version.

File: datavec-arrow/src/main/java/org/datavec/arrow/recordreader/ArrowWritableRecordBatch.java (hunk 4 above)
Comment (agibsonccc) [reply datavec-arrow/src/main/java/org/datavec/arrow/recordreader/ArrowWritableRecordBatch.java]: Have a commit id here?

Comment (agibsonccc) [reply datavec-arrow/src/main/java/org/datavec/arrow/recordreader/ArrowWritableRecordBatch.java]: Fixed.

File: datavec-arrow/src/main/java/org/datavec/arrow/recordreader/ArrowWritableRecordTimeSeriesBatch.java (hunk 6 above)
Comment (agibsonccc) [reply datavec-arrow/src/main/java/org/datavec/arrow/recordreader/ArrowWritableRecordTimeSeriesBatch.java]: Fixed.

**Pull request merged**